Sources/NIOPosix/MultiThreadedEventLoopGroup.swift

```diff
@@ -508,28 +508,38 @@ struct ErasedUnownedJob {
 
 @usableFromInline
 internal struct ScheduledTask {
+    @usableFromInline
+    enum Kind {
+        case task(task: () -> Void, failFn: (Error) -> Void)
+        case callback(any NIOScheduledCallbackHandler)
```

Comment:
One performance thought here. Currently we are storing an existential callback handler. However, we could just store the two closures itself which we can get while we are in the generic method. This way we would avoid calling through an existential on every scheduled callback task.

Reply:
The requirements for `protocol NIOScheduledCallbackHandler` are _generic_ functions. Is there a way I can store these as closures in the enum associated value, without making the `Kind` generic?

Reply:
I forgot to add the second half of this. Yes this is why I think we should go back to `any EventLoop`. I assume the perf benefit outweighs this. This is an assumption but the scheduling and running of tasks is probably hotter than whatever we do in their callback with the passed EL. @Lukasa WDYT?

Reply:
The functions being generic don't really matter here: we can store generic closures in the `Kind` type without promoting the generic to the `Kind` type itself. We won't get specialization, but that's fine.

So TL;DR: yes, changing `ScheduledEventLoop`'s representation to a pair of closures is probably the right thing to do.

Reply:
I spent some time working this and couldn't come up with a representation of this form that resulted in the amortized zero allocations we were looking for. Shall we shelve the suggestion on this thread? IIUC this is all internal anyway so we're not painting ourselves into a corner AFAICT?